File: nodejs/src/parsers/nginx_parser.ts
import * as fs from 'fs';
import { BaseParser, ParserData } from './base';

export class NginxParser extends BaseParser {
  formatName = 'nginx';
  fileExtensions = ['.log', '.access.log'];
  mimeTypes = ['text/plain'];

  private readonly LOG_PATTERN = /^(\S+) (\S+) (\S+) \[([^\]]+)\] "((\S+) (\S+) (\S+))" (\d+) (\d+) "([^"]*)" "([^"]*)"$/;

  supportsFormat(filePath: string): boolean {
    if (!filePath.toLowerCase().endsWith('.log') && !filePath.toLowerCase().endsWith('.access.log')) {
      return false;
    }

    try {
      const content = fs.readFileSync(filePath, 'utf-8');
      const firstLine = content.split('\n')[0].trim();
      return this.LOG_PATTERN.test(firstLine);
    } catch {
      return false;
    }
  }

  load(filePath: string): ParserData[] {
    const validation = this.validateFile(filePath);
    if (!validation.isValid) {
      console.error(`Error: ${validation.error}`);
      process.exit(1);
    }

    try {
      const content = fs.readFileSync(filePath, 'utf-8');
      const lines = content.trim().split('\n');
      const data: ParserData[] = [];

      for (let i = 0; i < lines.length; i++) {
        const line = lines[i].trim();
        if (!line) continue;

        const match = line.match(this.LOG_PATTERN);
        if (match) {
          const logEntry: ParserData = {
            remote_addr: match[1],
            remote_user: match[2],
            auth_user: match[3],
            time_local: this.parseNginxTimestamp(match[4]),
            request: match[5],
            method: match[6],
            path: match[7],
            protocol: match[8],
            status: parseInt(match[9], 10),
            body_bytes_sent: parseInt(match[10], 10),
            http_referer: match[11] === '-' ? null : match[11],
            http_user_agent: match[12] === '-' ? null : match[12],
          };

          data.push(logEntry);
        } else {
          console.warn(`Warning: Unable to parse line ${i + 1}`);
        }
      }

      return data;
    } catch (error) {
      console.error(`Error reading log file: ${error instanceof Error ? error.message : String(error)}`);
      process.exit(1);
    }
  }

  private parseNginxTimestamp(timestampStr: string): string {
    try {
      const parts = timestampStr.split(' ');
      const datePart = parts[0];
      
      const [day, month, yearTime] = datePart.split('/');
      const [year, time] = yearTime.split(':');
      const [hour, minute, second] = time.split(':');
      
      const monthMap: { [key: string]: number } = {
        'Jan': 0, 'Feb': 1, 'Mar': 2, 'Apr': 3, 'May': 4, 'Jun': 5,
        'Jul': 6, 'Aug': 7, 'Sep': 8, 'Oct': 9, 'Nov': 10, 'Dec': 11
      };
      
      const monthNum = monthMap[month];
      const date = new Date(
        parseInt(year, 10),
        monthNum,
        parseInt(day, 10),
        parseInt(hour, 10),
        parseInt(minute, 10),
        parseInt(second, 10)
      );
      
      return date.toISOString();
    } catch {
      return timestampStr;
    }
  }
}

File: parsers/nginx_parser.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Nginx access log parser (Combined format)
"""

import re
import sys
from typing import List, Dict, Any
from datetime import datetime
from .base import BaseParser


class NginxParser(BaseParser):
    """Parser for Nginx access logs (combined format)"""

    format_name = "nginx"
    file_extensions = ['.log', '.access.log']
    mime_types = ['text/plain']

    # Combined log format regex pattern
    # Example: 127.0.0.1 - - [10/Oct/2023:13:55:36 +0000] "GET /path HTTP/1.1" 200 1234 "http://referer" "Mozilla/5.0"
    LOG_PATTERN = re.compile(
        r'(?P<remote_addr>\S+) '                    # IP address
        r'(?P<remote_user>\S+) '                    # Remote user (usually -)
        r'(?P<auth_user>\S+) '                      # Auth user (usually -)
        r'\[(?P<time_local>[^\]]+)\] '              # Timestamp
        r'"(?P<request>(?P<method>\S+) '            # HTTP method
        r'(?P<path>\S+) '                           # Request path
        r'(?P<protocol>\S+))" '                     # HTTP protocol
        r'(?P<status>\d+) '                         # Status code
        r'(?P<body_bytes_sent>\d+) '                # Response size
        r'"(?P<http_referer>[^"]*)" '               # Referer
        r'"(?P<http_user_agent>[^"]*)"'             # User agent
    )

    def supports_format(self, file_path: str) -> bool:
        """Check if file looks like Nginx log"""
        if not file_path.lower().endswith(('.log', '.access.log')):
            return False

        # Try to parse first line to confirm format
        try:
            with open(file_path, 'r', encoding='utf-8') as f:
                first_line = f.readline().strip()
                return bool(self.LOG_PATTERN.match(first_line))
        except Exception:
            return False

    def load(self, file_path: str) -> List[Dict[str, Any]]:
        """Parse Nginx access log into structured data"""
        try:
            data = []
            with open(file_path, 'r', encoding='utf-8') as f:
                for line_num, line in enumerate(f, 1):
                    line = line.strip()
                    if not line:
                        continue

                    match = self.LOG_PATTERN.match(line)
                    if match:
                        log_entry = match.groupdict()

                        # Convert types
                        log_entry['status'] = int(log_entry['status'])
                        log_entry['body_bytes_sent'] = int(log_entry['body_bytes_sent'])

                        # Parse timestamp
                        try:
                            log_entry['time_local'] = self._parse_nginx_timestamp(
                                log_entry['time_local']
                            )
                        except ValueError:
                            pass  # Keep as string if parsing fails

                        # Handle empty strings
                        for key, value in log_entry.items():
                            if value == '-':
                                log_entry[key] = None

                        data.append(log_entry)
                    else:
                        print(f"警告: 无法解析第 {line_num} 行")

            return data

        except FileNotFoundError:
            print(f"错误: 文件 {file_path} 不存在")
            sys.exit(1)
        except Exception as e:
            print(f"读取日志文件错误: {e}")
            sys.exit(1)

    def _parse_nginx_timestamp(self, timestamp_str: str) -> str:
        """
        Parse Nginx timestamp format.
        Input: 10/Oct/2023:13:55:36 +0000
        Output: ISO format string
        """
        # Parse and convert to ISO format for easier SQL queries
        try:
            dt = datetime.strptime(timestamp_str.split(' ')[0], '%d/%b/%Y:%H:%M:%S')
            return dt.isoformat()
        except ValueError:
            return timestamp_str
